bindings/sundials4py/arkode/arkode.cpp

```diff
@@ -0,0 +1,244 @@
+ * Copyright (c) 2002-2013, Lawrence Livermore National Security.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ *----------------------------------------------------------------------------*/
+
+#include "sundials4py.hpp"
+
+#include <sundials/sundials_core.hpp>
+#include <sundials/sundials_stepper.hpp>
+
+#include <arkode/arkode.hpp>
+#include <arkode/arkode_arkstep.hpp>
+#include <arkode/arkode_butcher.h>
+#include <arkode/arkode_ls.h>
+#include <arkode/arkode_mristep.hpp>
+#include <arkode/arkode_sprkstep.hpp>
+
+#include "arkode/arkode_impl.h"
+#include "arkode_usersupplied.hpp"
+#include "sundials_adjointcheckpointscheme_impl.h"
+
+namespace nb = nanobind;
+using namespace sundials::experimental;
+
+namespace sundials4py {
+
+// Forward declarations of functions defined in other translation units
+void bind_arkode_erkstep(nb::module_& m);
+void bind_arkode_arkstep(nb::module_& m);
+void bind_arkode_sprkstep(nb::module_& m);
+void bind_arkode_lsrkstep(nb::module_& m);
+void bind_arkode_mristep(nb::module_& m);
+void bind_arkode_forcingstep(nb::module_& m);
+void bind_arkode_splittingstep(nb::module_& m);
+
+// ARKODE callback binding macros
+#define BIND_ARKODE_CALLBACK(NAME, FN_TYPE, MEMBER, WRAPPER, ...)       \
+  m.def(                                                                \
+    #NAME,                                                              \
+    [](void* ark_mem, std::function<std::remove_pointer_t<FN_TYPE>> fn) \
+    {                                                                   \
+      auto fn_table    = get_arkode_fn_table(ark_mem);                  \
+      fn_table->MEMBER = nb::cast(fn);                                  \
+      if (fn) { return NAME(ark_mem, &WRAPPER); }                       \
+      else { return NAME(ark_mem, nullptr); }                           \
+    },                                                                  \
+    __VA_ARGS__)
+
+#define BIND_ARKODE_CALLBACK2(NAME, FN_TYPE1, MEMBER1, WRAPPER1, FN_TYPE2, \
+                              MEMBER2, WRAPPER2, ...)                      \
+  m.def(                                                                   \
+    #NAME,                                                                 \
+    [](void* ark_mem, std::function<std::remove_pointer_t<FN_TYPE1>> fn1,  \
+       std::function<std::remove_pointer_t<FN_TYPE2>> fn2)                 \
+    {                                                                      \
+      auto fn_table     = get_arkode_fn_table(ark_mem);                    \
+      fn_table->MEMBER1 = nb::cast(fn1);                                   \
+      fn_table->MEMBER2 = nb::cast(fn2);                                   \
+      if (fn1) { return NAME(ark_mem, &WRAPPER1, &WRAPPER2); }             \
+      else { return NAME(ark_mem, nullptr, &WRAPPER2); }                   \
```

Comment:
Is `fn2` always non-null or should this be
```suggestion
      if (fn1 && fn2) { return NAME(ark_mem, &WRAPPER1, &WRAPPER2); }      \
      else if (fn1) { return NAME(ark_mem, &WRAPPER1, nullptr); }          \
      else if (fn2) { return NAME(ark_mem, nullptr, &WRAPPER2); }          \
      else { return NAME(ark_mem, nullptr, nullptr); }                     \
```